scipy/stats/_quantile.py

```diff
@@ -0,0 +1,268 @@
+    ``linear`` (default)            7               ``1 - p``
+    ``median_unbiased``             8               ``p/3 + 1/3``
+    ``normal_unbiased``             9               ``p/4 + 3/8``
+    =============================== =============== ===============
+
+    Note that indices ``j`` and ``j + 1`` are clipped to the range ``0`` to
+    ``n - 1`` when the results of the formula would be outside the allowed
+    range of non-negative indices. The ``- 1`` in the formulas for ``j`` and
+    ``g`` accounts for Python's 0-based indexing.
+
+    The table above includes only the estimators from H&F that are continuous
+    functions of probability `p` (estimators 4-9). NumPy also provides the
+    three discontinuous estimators from H&F (estimators 1-3), where ``j`` is
+    defined as above, ``m`` is defined as follows, and ``g`` is a function
+    of the real-valued ``index = p*n + m - 1`` and ``j``.
+
+    1. ``inverted_cdf``: ``m = 0`` and ``g = int(index - j > 0)``
+    2. ``averaged_inverted_cdf``: ``m = 0`` and
+       ``g = (1 + int(index - j > 0)) / 2``
+    3. ``closest_observation``: ``m = -1/2`` and
+       ``g = 1 - int((index == j) & (j%2 == 1))``
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> x = np.asarray([[10, 8, 7, 5, 4],
+    ...                 [0, 1, 2, 3, 5]])
+
+    Take the median along the last axis.
+
+    >>> stats.quantile(x, 0.5, axis=-1)
+    array([7.,  2.])
+
+    Take a different quantile along each axis.
+
+    >>> stats.quantile(x, [[0.25], [0.75]], axis=-1, keepdims=True)
+    array([[5.],
+           [3.]])
+
+    Take multiple quantiles along each axis.
+
+    >>> stats.quantile(x, [0.25, 0.75], axis=-1)
+    array([[5., 8.],
+           [1., 3.]])
+
+    References
+    ----------
+    .. [1] R. J. Hyndman and Y. Fan,
+       "Sample quantiles in statistical packages,"
+       The American Statistician, 50(4), pp. 361-365, 1996
+
+    """
+    # Input validation / standardization
+
+    temp = _quantile_iv(x, p, method, axis, nan_policy, keepdims)
+    y, p, method, axis, nan_policy, keepdims, n, axis_none, ndim, p_mask, xp = temp
+
+    # Algorithm
+    methods = dict(interpolated_inverted_cdf=0, hazen=0.5, weibull=p, linear=1 - p,
+                   median_unbiased=p/3 + 1/3, normal_unbiased=p/4 + 3/8)
+    m = methods[method]
+    jg = p*n + m - 1
+    j = jg // 1
+    g = jg % 1
+    j = xp.clip(j, 0., n - 1)
+    jp1 = xp.clip(j + 1, 0., n - 1)
+    j = xp.astype(j, xp.int64)
+    jp1 = xp.astype(jp1, xp.int64)
+
+    res = ((1 - g) * xp_take_along_axis(y, j, axis=-1)
+           + g * xp_take_along_axis(y, jp1, axis=-1))
+    res[p_mask] = xp.nan
+
+    # Reshape per axis/keepdims
+    if axis_none and keepdims:
+        shape = (1,)*(ndim - 1) + res.shape
+        res = xp.reshape(res, shape)
+        axis = -1
+
+    res = xp.moveaxis(res, -1, axis)
+
+    if not keepdims:
+        res = xp.squeeze(res, axis=axis)
+
+    return res[()]
```

Comment:
This won't work on array-api-strict, correct? In general, this is not compliant unless res is strictly 1D

Reply:
Then we will do the usual `res[()] if res is None else res`. Surely there will be other unfortunate verbosities I forgot, and the test suite will let us know.